server/socketRoomManager.ts

```diff
@@ -0,0 +1,78 @@
+// Description: This file contains the logic to handle allocation of players
+// into rooms and recursive reduction of rooms until a winner is determined.
+
+// Contains sample logic. Needs to be refactored and integrated with the duel logic.
+
+import io from "./server";
+import Player from "./model/actors/player";
+
+// Function to handle allocation of players and recursive reduction of rooms
+export const handleRoomAllocation = async (players: Player[]) => {
+  const winners: Player[] = [];
+
+  const numPlayers = players.length;
+  const numRooms = Math.ceil(numPlayers / 2); // Each room will have 2 players
+
+  // Divide players into groups for each room
+  // tournamentManager should be responsible for this logic
+  const playerGroups: Player[][] = [];
+  for (let i = 0; i < numRooms; i++) {
+    const startIndex = i * 2;
+    const endIndex = Math.min(startIndex + 2, numPlayers);
+    const group = players.slice(startIndex, endIndex);
+    playerGroups.push(group);
+  }
+
+  // Iterate through each group of players
+  for (const group of playerGroups) {
+    // Generate a unique room name
+    const roomName = generateUniqueRoomName();
+
+    // Allocate players to their respective room
+    for (const player of group) {
+      const socket = io.sockets.sockets.get(player.socketId);
+      if (socket) {
+        await socket.join(roomName); // Join the socket to the room
+      }
+    }
+
+    // Emit "CHOOSE_PLAYER_MOVE" event to each room
+    io.to(roomName).emit("CHOOSE_PLAYER_MOVE");
+  }
+
+  // Wait for results from each room
+  await waitForResults(playerGroups, winners);
+
+  // If more than one room remains, recursively handle allocation
+  if (playerGroups.length > 1) {
+    await handleRoomAllocation(winners);
+  } else {
+    // Implement logic to determine the winner
+    console.log("Game ended");
+    const winningPlayer = winners[0]; // Assuming there is only one winner
+    const winningSocket = io.sockets.sockets.get(winningPlayer.socketId);
+    if (winningSocket) {
+      // Emit a socket message to the winning player
+      winningSocket.emit("GAME_WINNER", { winner: winningPlayer });
+    }
+    return;
+  }
+};
+
+// Generate a random alphanumeric room name
+function generateUniqueRoomName(): string {
```

Comment:
Rather then generating a random unique room name (which this isnt technically as you could potentially get the same random number eventually right?) you could instead just do some like "GAME_${gameCode}&ROOM_${roomNumber}".